crates/ruff_linter/src/noqa.rs

```diff
@@ -418,182 +300,408 @@ impl<'a> FileNoqaDirectives<'a> {
+                range: TextRange::new(self.offset, last.range.end()),
+                codes: self.codes,
+            }),
+        })
     }
 
-    /// Lex a specific character, or return `None` if the character is not the first character in
-    /// the line.
-    #[inline]
-    fn lex_char(line: &str, c: char) -> Option<&str> {
-        let mut chars = line.chars();
-        if chars.next() == Some(c) {
-            Some(chars.as_str())
-        } else {
-            None
+    fn lex_codes(&mut self) -> Result<(), LexicalError> {
+        // SAFETY: Every call to `lex_code` advances the cursor at least once.
+        while !self.cursor.is_eof() {
+            self.lex_code()?;
         }
+        Ok(())
     }
 
-    /// Lex the "flake8" prefix of a `noqa` directive.
-    #[inline]
-    fn lex_flake8(line: &str) -> Option<&str> {
-        line.strip_prefix("flake8")
+    fn lex_code(&mut self) -> Result<(), LexicalError> {
+        self.cursor.start_token();
+        self.cursor.eat_while(|chr| chr.is_ascii_whitespace());
+
+        // Ex) # noqa: F401,    ,F841
+        //                  ^^^^
+        if self.cursor.first() == ',' {
+            self.warnings.push(LexicalWarning::MissingItem(
+                self.token_range().add(self.offset),
+            ));
+            self.cursor.eat_char(',');
+            return Ok(());
+        }
+
+        // Reset start of token so it does not include whitespace
+        self.cursor.start_token();
+        match self.cursor.bump() {
+            // Ex) # noqa: F401
+            //             ^
+            Some(c) if c.is_ascii_uppercase() => {
+                self.cursor.eat_while(|chr| chr.is_ascii_uppercase());
+                if !self.cursor.eat_if(|c| c.is_ascii_digit()) {
+                    // Fail hard if we're already attempting
+                    // to lex squashed codes, e.g. `F401Fabc`
+                    if self.missing_delimiter {
+                        return Err(LexicalError::InvalidCodeSuffix);
+                    }
+                    // Otherwise we've reached the first invalid code,
+                    // so it could be a comment and we just stop parsing.
+                    // Ex) #noqa: F401 A comment
+                    //       we're here^
+                    self.cursor.skip_bytes(self.cursor.as_str().len());
+                    return Ok(());
+                }
+                self.cursor.eat_while(|chr| chr.is_ascii_digit());
+                self.push_code();
+
+                if self.cursor.is_eof() {
+                    return Ok(());
+                }
+
+                self.missing_delimiter = match self.cursor.first() {
+                    ',' => {
+                        self.cursor.eat_char(',');
+                        false
+                    }
+
+                    // Whitespace could be a delimiter or the end of the `noqa`.
+                    // If it's the end, then non-ascii whitespace is allowed
```

Comment:
Can we add an example here? E.g. what about `noqa: RUF100 RUF200`? It's also not clear to me what you mean by non-ascii whitespace